File: 3week/근우/2. coin-change.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        NOT_VISIT = 2e9
        VISIT_BUT_NOT_FOUND = 2e9 + 1
        coins = list(filter(lambda x: x <= amount, coins))
        coins.sort(reverse=True)
        cache = [NOT_VISIT for _ in range(amount + 1)]

        def _dp(cur):
            if cache[cur] != NOT_VISIT:
                return cache[cur]
            elif cur == amount:
                return 0
            res = VISIT_BUT_NOT_FOUND
            for c in coins:
                next_cur = cur + c
                if next_cur <= amount:
                    res = min(res, 1 + _dp(next_cur))
            cache[cur] = res
            return res

        res = _dp(0)
        return res if res != VISIT_BUT_NOT_FOUND else -1

File: 3week/남규/2. coin change.cpp
class Solution {
public:
    int dfs(vector<int>& coins, vector<vector<int>>& dp, int remain, int index){
        int N = coins.size();
        if(remain == 0) return 0;
        if(index >= N) return INT_MAX;
        
        int &ret = dp[remain][index];
        if(ret != -1) return ret;
        ret = INT_MAX;
        
        if(remain - coins[index] >= 0){
            int cand = dfs(coins, dp, remain - coins[index], index);
            cand = cand == INT_MAX ? cand : cand + 1;
            ret = min(ret, cand);
        }
        ret = min(ret, dfs(coins, dp, remain, index + 1));
        return ret;              
    }
    
    int coinChange(vector<int>& coins, int amount) {
        if(amount == 0) return 0;
        int N = coins.size();
        vector<vector<int>> dp(amount + 1, vector<int>(N, -1));
        int ans = dfs(coins, dp, amount, 0);
        return ans == INT_MAX ? -1 : ans;
    }
};
